# Text processing with regex {#regex}

```{r setup-stringrr, include=FALSE}
options(digits = 3)
knitr::opts_knit$set(cache = TRUE,
                     tidy = TRUE,
                     tidy.opts = list(blank = FALSE, width.cutoff = 60))
library(stringr)
```

## Regex syntax

### Why regexes?

This chapter deals with processing data in textual form: character data.  
When working with text, you really need to be able to work with regular expressions. That is why these are dealt with first, together with the base R regex functions. After that the functions from the `stringr` package are discussed.


It is easy enough to look for the word "Chimpanzee" in a vector containing animal species names:

```{r regex-data-1}
animals = c("Chimpanzee", "Cow", "Camel")
animals == "Chimpanzee"
```

but what are you going to do if there are multiple variants of the word you are looking for? This?

```{r regex-demo-1}
animals = c("Chimpanzee", "Chimp", "chimpanzee", "Camel")
animals == "Chimpanzee" | animals == "Chimp" | animals == "chimpanzee"
```

The solution here is not using literals, but to describe **_patterns_**.

Look at the above example. How would you describe a pattern that would correctly identify all Chimpanzee occurrences?

Is you pattern something like this?

_A letter C in upper-or lower case followed by 'himp' followed by nothing or 'anzee'_

In programming we use **_regular expressions_** or **_RegEx_** to describe such a pattern in a formal concise way:

`[Cc]himp(anzee)?`

And to apply such a pattern in R, we use one of several functions dedicated for this task. Here is one, `grepl()`, which returns `TRUE` if the regex matched the vector element.

```{r regex-demo-2}
grepl("[Cc]himp(anzee)?", animals)
```


:::definitions
Pattern matching is the process of finding, locating, extracting and replacing patterns in character data that usually cannot be literally described.
:::


#### Base functions using regex {-}

There are several base R functions dedicated to finding patters in character data. 
They differ in intent and output. Later, the `stringr` counterparts will be discussed.

- **finding** Does an element contain a pattern (TRUE/FALSE)? `grepl(pattern, string)`
- **locating** Which elements contain a pattern (INDEX)? `grep(pattern, string)`
- **extracting** Get the content of matching elements `grep(pattern, string, value = TRUE)`
- **replace** Replace the first occurrence of the pattern `sub(pattern, replacement, string)`
- **replace all** Replace all occurrences of the pattern `gsub(pattern, replacement, string)`

Note that the `stringr` package from the tidyverse has many user-friendly functions in this field as well. Two of them will be dealt with in the exercises.


### Regex components

A regular expression can be build out of any combination of

- **character sequences** - Literal character sequences such as 'chimp'
- **character classes** - A listing of possibilities for a single position.
    - Between brackets: `[adgk]` means 'a' or 'd' or 'g' or 'k'.
    - Use a hyphen to create a series: `[3-9]` means digits 3 through 9 and `[a-zA-Z]` means all alphabet characters.
    - Negate using `^`. `[^adgk]` means anything _but_ a, d, g or k.
    - A special case is the dot `.`: any character matches.
    - Many special character classes exist (digits, whitespaces etc). They are discussed in a later paragraph.
- **alternatives** - Are defined by the pipe symbol `|`: "OR"
- **quantifiers** - How many times the preceding block should occur. See next paragraph.
- **anchors** - `^` means matching at the start of a string. `$` means at the end.

An excellent cheat sheet from the RStudio website is also included
<a href="figures/RegExCheatsheet.pdf" target="_blank">here</a>



#### Quantifiers {-}

Use quantifiers to specify how many times a character or series of characters should occur.

- **`{n}`**: exactly `n` times
- **`{n, }`**: at least `n` times
- **`{ ,n}`**: at most `n` times
- **`{n, m}`**: at least `n` and at most `m` times.
- **`*`**: 0 or more times; same as `{0, }`
- **`+`**: 1 or more times; same as `{1, }`
- **`?`**: 0 or 1 time; same as `{0, 1}`


#### Anchoring {-}

Using anchoring, you can make sure the string is not longer than you explicitly state:

```{r regex-anchoring}
dates <- c("15/2/2019", "15-2-2019", "15-02-2019", "015/2/20191", "15/2/20191")
dateRegex <- "^[0-9]{2}[/-][0-9]{1,2}[/-][0-9]{4}$"
grep(pattern = dateRegex, x = dates, value = TRUE)
```

Now the date matching is correct.

#### Metacharacters: Special character classes {-}

Since patterns such as `[0-9]` occur so frequently, they have dedicated character classes such as `[[:digit:]]`. The most important other ones are

- **digits** `[[:digit:]]` or `\\d`: equivalent to `[0-9]`
- **alphabet characters** `[[:alpha:]]`: equivalent to `[a-zA-Z]`
- **lowercase characters** `[[:lower:]]`: equivalent to `[a-z]`
- **uppercase characters** `[[:upper:]]`: equivalent to `[A-Z]`
- **whitespace characters** `[[:space:]]` or `\\s`: Space, tab, vertical tab, newline, form feed, carriage return
- **punctuation characters** `[[:punct:]]`: One of !"#$%&â€™()*+,-./:;<=>?@[]^_`{|}~

(have a look at the cheat sheet for all)

Here is the same example, this time using these predefined character classes

```{r regex-character-classes-1}
dates <- c("15/2/2019", "15-2-2019", "15-02-2019", "15022019", "15/2/20191")
dateRegex <- "[[:digit:]]{2}[/-]\\d{1,2}[/-]\\d{4}"
grep(pattern = dateRegex, x = dates, value = TRUE)
```


#### Alternatives {-}

To apply two alternative choices for a single regex element you use the pipe symbol `|`. You can us parentheses `(foo[])` to fence alternatives off.

```{r regex-alternatives, warning=FALSE}
column_names <- c("Subject", "Age", "T0_creatine", "T0_calcium", "T1_creatine", "T1_calcium") 
grep(pattern = "T[01]_(creatine|calcium)", x = column_names, value = TRUE)
```


### Some examples

#### Restriction enzymes {-}

This is the recognition sequence for the _HincII_ restriction endonuclease:

```
5'-GTYRAC-3'
3'-CARYTG-5'
```

Before reading on: how would you define a regular expression that is precisely describes this recognition sequence?

Molecular biology sequence ambiguity codes can be found
<a href="data/ambiguity-codes.html" target="_blank">here</a>

```{r hinc2-pattern}
HincII_rs <- "GT[CT][AG]AC"
sequences <- c("GTCAAC",
               "GTCGAC",
               "GTTGAC",
               "aGTTAACa",
               "GTGCAC")
grep(pattern = HincII_rs, x = sequences, value = TRUE)
```

#### Dutch dates {-}

Here are some Dutch dates, in different accepted formats. The last two are not a correct notation.
Create a RegEx that will determine whether an element contains a Dutch date string.

```{r regex-dutch-dates}
dates <- c("15/2/2019", "15-2-2019", "15-02-2019", "015/2/20191", "15/2/20191")
dateRegex <- "[0-9]{2}[/-][0-9]{1,2}[/-][0-9]{4}"
grep(pattern = dateRegex, x = dates, value = TRUE)
```

Why were the last two matched?
Because the pattern _is there_, albeit embedded in a longer string.
We have to **_anchor_** the pattern to be more specific.

#### Exercise: Postal codes {-}

Here are some Dutch zip (postal) codes, in different accepted formats. The last two are not a correct notation.
Can you create a RegEx that will determine whether an element contains a Dutch zip code?

```{r regex-postal-codes}
zips <- c("1234 AA", "2345-BB", "3456CC", "4567 dd", "56789aa", "6789a_")
zips
```


#### Exercise: Prosite patterns {-}

<a href="https://prosite.expasy.org/" target="_blank">Prosite</a> is a database of amino acid sequence motifs. One of them is the Histidine Triad profile (PDOC00694).

```
[NQAR]-x(4)-[GSAVY]-x-[QFLPA]-x-[LIVMY]-x-[HWYRQ]-
[LIVMFYST]-H-[LIVMFT]-H-[LIVMF]-[LIVMFPT]-[PSGAWN]
```

- Write this down as a RegEx
- Was that efficient? Using the `gsub()` function, can you convert it in a RegEx using code? It may take several iterations. Was that efficient?
- Next, use an appropriate function to find if, and where, this pattern is located within the sequences in file `data/hit_proteins.txt` (<a href="data/hit_proteins.txt" target="_blank">here</a>)

Amino Acid codes and Prosite pattern encoding can be found
<a href="data/ambiguity-codes.html" target="_blank">here</a>


## The `stringr` package

This is the last presentation in the tidyverse series. It revolves around processing textual data: finding, extracting, and replacing patterns. Central to this task is **_pattern matching_** using regular expressions. Pattern matching is the process of finding, locating, extracting and replacing patterns in character data that usually cannot be literally described. **_Regular expression syntax_** is the language in which patterns are described in a wide range of programming languages, including R. 

This topic has been dealt with in an introductory manner previously (course DAVuR1). And is repeated and expanded here. Instead of the base R functions we now switch to the `stringr` package.

As all packages from the tidyverse, `stringr` has many many functions (type `help(package = "stringr")` to see which). this package has a great [cheat sheet](graphics/stringr.pdf) as well.

Here, a few of them will be reviewed.

### A few remarks on "locale"

Many functions of the tidyverse packages related to time and text (and currency) accept arguments specifying the **_locale_**.
The locale is a container for all location-specific display of information.  
Think  

- Character set of the language
- Time zone, Daylight savings time
- Thousands separator and decimal symbol
- Currency symbol

Dealing with locales is a big challenge indeed for any programming language. However, since this is only an introductory course we will stick to US English and work with the current locale for times only.
This note is to make you aware of the concept so that you remember this when the appropriate time comes.



### The `stringr` essentials


#### Case conversion {-}

These functions all change the capitalization of (some of) the word characters of an input string. They all ignore non-word characters such as punctuation and other symbols.

- `str_to_upper()` converts all word characters to uppercase
- `str_to_lower()` converts all word characters to lowercase
- `str_to_title()` capitalizes all first characters of words
- `str_to_sentence()` capitalizes the first character in the string, not after every period


```{r rstringr-case-conv}
sntc <- "the path of the righteous man is beset on all sides by the iniquities of the selfish,  and the tyranny of evil men. --quote from?"

str_to_title(sntc)
str_to_sentence(sntc)
```

#### Split, join and substring {-}

Combining two vectors into one, one vector into one, or doing the reverse: splitting. These are all string-based operation that are carried out in scripting quite often.

Here are some joining operations, using `str_c()`:

```{r stringr-join}
l1 <- letters[1:5]
l2 <- letters[6:10]

str_c(l1, collapse = "=")
str_c(l1, l2, sep = "+")
str_c(l1, l2, sep = "+", collapse = "=")
```

When you want to combine variables and text `str_glue()` comes in handy:

```{r stringr-glue}
str_glue("The value of pi is {pi} and the first month of the year is {month.name[1]}")
```

This is a more friendly approach than with `paste()`.

**Splitting** is slightly more tricky since it accepts a regex pattern as split argument. For instance, you can get the words of a sentence by splitting like this:

```{r stringr-split, results = "hold"}
words <- str_split(sntc, "([[:punct:]]|[[:space:]])+")
words
##alternative
#str_split(sntc, "[^a-zA-Z]+")
```

There are two ways to get parts of character strings, or **_substrings_**. The first is by index. You can omit both `start` and `end` arguments; they will default to start and end of the string, respectively.

```{r stringr-sub-1}
nucs <- c("Adenine", "Guanine", "Cytosine", "Thymine")
str_sub(nucs, end = 3)
```

You can even use this function to change the substring that is removed
```{r stringr-sub-2}
str_sub(nucs, start = 4) <- "......"
nucs
```

This does not work with literals! The following chunk gives and error:

```{r stringr-sub-3, error = TRUE}
str_sub(c("Adenine", "Guanine", "Cytosine", "Thymine"), start = 4) <- "......"
```


#### Matching {-}

When you **match** a pattern to a string, you usually want to know **_if_** it is there, **which** elements have it, **_where_** it is located in those elements or **_how often_** it is present. For each of these question there is a dedicated function:

```{r stringr-demo-data, include = FALSE}
fruits <- c("Banana", "Apple", "Orange", "Cherry")
```

- `str_detect(string, pattern)` detects the presence of a pattern match in a string. 
    ```{r str-detect}
    str_detect(fruits, "[Aa]")
    ```

- `str_subset(string, pattern)` returns only the strings that contain a pattern match
    ```{r str-subset}
    str_subset(fruits, "[Aa]")
    ```

- `str_which(string, pattern)` finds the indexes of strings that contain a pattern match.
    ```{r str-which}
    str_which(fruits, "[Aa]")
    ```

- `str_count(string, pattern)` counts the number of matches in a string.
    ```{r str-count}
    str_count(fruits, "[Aa]")
    ```

- `str_locate(string, pattern)` and `str_locate_all(string, pattern)` locate the positions of pattern matches in a string

    ```{r str-locate-all}
    str_locate_all(fruits, "[Aa]")
    ```

#### Extracting and replacing {-}

If you want to obtain the character sequences matching your pattern you can use the `str_extract()` and `str_extract_all()` functions:

```{r str-extract-all}
str_extract_all(fruits, "an")
```

Finally, replacing occurrences of a pattern is carried out using `str_replace()` or `str_replace_all()`.

```{r str-replace-all}
str_replace_all(fruits, "an", "..")
```





